core/src/forwarding_stage.rs

```diff
@@ -0,0 +1,618 @@
+            let packet = self
+                .packet_container
+                .packets
+                .get(priority_index.index)
+                .expect("packet exists");
+
+            // If it exceeds our data-budget, drop.
+            if !self.data_budget.take(packet.meta().size) {
+                self.metrics.votes_dropped_on_data_budget +=
+                    usize::from(packet.meta().is_simple_vote_tx());
+                self.metrics.non_votes_dropped_on_data_budget +=
+                    usize::from(!packet.meta().is_simple_vote_tx());
+                self.packet_container.packets.remove(priority_index.index);
+                continue;
+            }
+
+            let packet_data_vec = packet.data(..).expect("packet has data").to_vec();
+
+            if packet.meta().is_simple_vote_tx() {
+                vote_batch.push(packet_data_vec);
+                if vote_batch.len() == vote_batch.capacity() {
+                    self.send_vote_batch(tpu_vote, &mut vote_batch);
+                }
+            } else {
+                non_vote_batch.push((packet_data_vec, tpu));
+                if non_vote_batch.len() == non_vote_batch.capacity() {
+                    self.send_non_vote_batch(&mut non_vote_batch);
+                }
+            }
+        }
+
+        // Send out remaining packets
+        if !vote_batch.is_empty() {
+            self.metrics.votes_forwarded += vote_batch.len();
+            self.send_vote_batch(tpu_vote, &mut vote_batch);
+        }
+        if !non_vote_batch.is_empty() {
+            self.metrics.non_votes_forwarded += non_vote_batch.len();
+            self.send_non_vote_batch(&mut non_vote_batch);
+        }
+    }
+
+    /// Re-fill the data budget if enough time has passed
+    fn refresh_data_budget(&self) {
+        const INTERVAL_MS: u64 = 100;
+        // 12 MB outbound limit per second
+        const MAX_BYTES_PER_SECOND: usize = 12_000_000;
+        const MAX_BYTES_PER_INTERVAL: usize = MAX_BYTES_PER_SECOND * INTERVAL_MS as usize / 1000;
+        const MAX_BYTES_BUDGET: usize = MAX_BYTES_PER_INTERVAL * 5;
+        self.data_budget.update(INTERVAL_MS, |bytes| {
+            std::cmp::min(
+                bytes.saturating_add(MAX_BYTES_PER_INTERVAL),
+                MAX_BYTES_BUDGET,
+            )
+        });
+    }
+
+    fn send_vote_batch(&self, addr: SocketAddr, vote_batch: &mut Vec<Vec<u8>>) {
+        let conn = self.connection_cache.get_connection(&addr);
+        let mut batch = Vec::with_capacity(FORWARD_BATCH_SIZE);
+        core::mem::swap(&mut batch, vote_batch);
+        let _res = conn.send_data_batch_async(batch);
+    }
+
+    fn send_non_vote_batch(&self, non_vote_batch: &mut Vec<(Vec<u8>, SocketAddr)>) {
+        let mut batch = Vec::with_capacity(FORWARD_BATCH_SIZE);
+        core::mem::swap(&mut batch, non_vote_batch);
+        let _res = batch_send(&self.udp_socket, &batch);
+    }
+}
+
+struct PacketContainer {
```

Comment:
This should probably be its own file with unittests etc

Reply:
Module separation: fb06eef5f0d2e0762c0cea0cd1ec8c6432409694
Added tests: 24630c8d58ae96785a718ea782d6698c13f162d6